File: python/core/eventbus.py
"""
EventBus — Sóng hấp dẫn (Gravitational Waves).
Giao tiếp gián tiếp giữa modules. Principle #5.
"""
from typing import Any, Callable


class EventBus:
    """In-process event bus. String-keyed, type-safe via convention."""

    def __init__(self) -> None:
        self._handlers: dict[str, list[Callable]] = {}

    def publish(self, event_type: str, event: Any) -> None:
        """Publish event to all subscribers."""
        for handler in list(self._handlers.get(event_type, [])):
            handler(event)

    def subscribe(self, event_type: str, handler: Callable) -> None:
        """Subscribe to event type."""
        self._handlers.setdefault(event_type, []).append(handler)

    def unsubscribe(self, event_type: str, handler: Callable) -> None:
        """Remove subscription."""
        handlers = self._handlers.get(event_type)
        if handlers and handler in handlers:
            handlers.remove(handler)

    @property
    def type_count(self) -> int:
        return len(self._handlers)

    @property
    def handler_count(self) -> int:
        return sum(len(h) for h in self._handlers.values())

File: typescript/src/core/EventBus.ts
/**
 * EventBus — Sóng hấp dẫn (Gravitational Waves).
 * Giao tiếp gián tiếp giữa modules. Principle #5.
 */
export interface IEventBus {
  publish<T>(eventType: string, event: T): void;
  subscribe<T>(eventType: string, handler: (event: T) => void): void;
  unsubscribe<T>(eventType: string, handler: (event: T) => void): void;
}

export class EventBus implements IEventBus {
  private readonly handlers = new Map<string, Array<(event: any) => void>>();

  publish<T>(eventType: string, event: T): void {
    const handlers = this.handlers.get(eventType);
    if (!handlers) return;
    for (const handler of [...handlers]) handler(event);
  }

  subscribe<T>(eventType: string, handler: (event: T) => void): void {
    if (!this.handlers.has(eventType)) {
      this.handlers.set(eventType, []);
    }
    this.handlers.get(eventType)!.push(handler);
  }

  unsubscribe<T>(eventType: string, handler: (event: T) => void): void {
    const handlers = this.handlers.get(eventType);
    if (!handlers) return;
    const idx = handlers.indexOf(handler);
    if (idx >= 0) handlers.splice(idx, 1);
  }

  get typeCount(): number { return this.handlers.size; }
  get handlerCount(): number {
    let count = 0;
    for (const h of this.handlers.values()) count += h.length;
    return count;
  }
}
